go-controller/pkg/node/node_ip_handler_linux.go

```diff
@@ -0,0 +1,149 @@
+// +build linux
+
+package node
+
+import (
+	"net"
+	"sync"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/kube"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+
+	"github.com/vishvananda/netlink"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+	utilnet "k8s.io/utils/net"
+)
+
+type addressManager struct {
+	addresses      sets.String
+	nodeAnnotator  kube.Annotator
+	mgmtPortConfig *managementPortConfig
+	mu             sync.Mutex
+}
+
+// initializes a new address manager which will hold all the IPs on a node
+func newAddressManager(nodeAnnotator kube.Annotator, config *managementPortConfig) *addressManager {
+	mgr := &addressManager{
+		addresses:      sets.NewString(),
+		nodeAnnotator:  nodeAnnotator,
+		mgmtPortConfig: config,
+		mu:             sync.Mutex{},
+	}
+
+	klog.Info("Address manager created with IPs: %s", mgr.addresses.List())
+	return mgr
+}
+
+// updates the address manager with a new IP
+// returns true if there was an update
+func (c *addressManager) addAddr(ip net.IP) bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if !c.addresses.Has(ip.String()) && c.isValidNodeIP(ip) {
+		klog.V(5).Infof("Adding IP: %s, to node IP manager", ip)
+		c.addresses.Insert(ip.String())
+		return true
+	}
+
+	return false
+}
+
+// removes IP from address manager
+// returns true if there was an update
+func (c *addressManager) delAddr(ip net.IP) bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.addresses.Has(ip.String()) && c.isValidNodeIP(ip) {
+		klog.V(5).Infof("Removing IP: %s, from node IP manager", ip)
+		c.addresses.Delete(ip.String())
+		return true
+	}
+
+	return false
+}
+
+func (c *addressManager) Run(stopChan <-chan struct{}) {
+	addrChan := make(chan netlink.AddrUpdate)
+	if err := netlink.AddrSubscribe(addrChan, stopChan); err != nil {
+		klog.Fatalf("Unable to run Address Manager, error during netlink subscribe: %v", err)
+	}
+	go func() {
+		for {
+			select {
+			case a := <-addrChan:
+				if a.NewAddr {
+					if c.addAddr(a.LinkAddress.IP) {
+						if err := util.SetNodeHostAddresses(c.nodeAnnotator, c.addresses); err != nil {
+							klog.Errorf("Failed to set node annotations: %v", err)
+							continue
+						}
+						if err := c.nodeAnnotator.Run(); err != nil {
+							klog.Errorf("Failed to set node annotations: %v", err)
+						}
+					}
+				} else {
+					if c.delAddr(a.LinkAddress.IP) {
+						if err := util.SetNodeHostAddresses(c.nodeAnnotator, c.addresses); err != nil {
+							klog.Errorf("Failed to set node annotations: %v", err)
+							continue
+						}
+						if err := c.nodeAnnotator.Run(); err != nil {
```

Comment:
@trozet why not move the `if err := c.nodeAnnotator.Run()` outside the if/else? Same code run for both add/del.